<h1> Actividad 1 - Pilas y colas </h1>
Carmen Witsman García

En esta actividad, implementaremos una pila dinámica, con estructura LIFO (Last-in, First-out), de las siguientes formas:

- Con la técnica de construcción de lista enlazada: Clase `LinkedStack`
- Con soporte de Lista de Python: Clase `ListStack`
- Con soporte de estructura DeQue: Clase `DqueStack`
- Con soporte de estructura LifoQueue: Clase `LifoQueueStack`

Todas las clases que crearemos tendrán los siguientes métodos:

- `__init__` - Método constructor con el que inicializaremos la pila vacía
- `push(e)` - Apila un elemento (e) al inicio de la pila
- `pop()` - Extrae y devuelve el primer elemento de la pila
- `peek()` - Devuelve la cima de la pila (el primer elemento)
- `empty()` - Comprueba si la pila está vacía
- `clear()` - Elimina todos los elementos de la pila

Comentaremos para cada clase la forma en la que implementamos dichos métodos, y mostraremos ejemplos de uso de cada uno de ellos.

Al final, realizaremos una comparativa de estos 4 métodos a nivel de rendimiento y facilidad de uso y de comprensión.


<h2> 1. Pila dinámica con lista enlazada </h2>

In [10]:
# Primero, definimos la clase Nodo para implentarla en LinkedStack

class Node:
    def __init__(self, data=None):
        self.data = data
        self.next = None
    def __str__(self):              # Devuelve la representeación en string del atributo "data"
        return str(self.data)  

In [11]:
# Creamos la clase

class LinkedStack:

    def __init__(self):
        """
        Método constructor.
        """
        # Pila vacía: define el puntero first a None
        self.first = None
        
    def __str__(self):
        """
        Mostramos la pila en formato string.
        """
        nodo = self.first
        llist = []                  # Usa una lista vacía para almacenar los elementos de la pila
        while nodo:
            llist.append(str(nodo)) # Añade uno por uno los nodos (elementos de la pila)
            nodo = nodo.next        # Actualiza el puntero
        if len(llist) > 0:
            return str(llist)       # Si la lista no está vacía, devuelve la lista en formato string
        return "Pila vacía"         # Si no, devuelve "Pila vacía"


    def push(self, e):
        """
        Apila un elemento al principio de la pila.
        """
        # Actualizamos nodo para que apunte al elemento introducido
        nodo = Node(e)

        if(self.first is None):      # La pila está vacía
            self.first = nodo        # El puntero "first" apunta a nodo (el elemento)

        else:                        # La pila no está vacía 
            nodo.next = self.first   # Añade el nodo al principio
            self.first = nodo        # Actualiza el puntero "first" a nodo
            
              
    def pop(self):
        """
        Borra el elemento del principio de la pila y lo devuelve.
        """
        try:
            # Elimina el primer elemento y guarda su valor
            e1 = self.first

            # Si la lista tiene solo un elemento, elimina el nodo "first"
            if self.first.next == None:
                self.first = None
            else:
                # Si la pila tiene más elementos, actualiza el puntero "first" al siguiente nodo 
                self.first = self.first.next

            return str(e1) # Devuelve el elemento eliminado
        
        except IndexError: # Si la pila está vacía, lanza una excepción
            return "La pila está vacía"
    
          
                
    def peek(self):
        """
        Devuelve la cima de la pila (el 1º elemento).
        """
        try:
            # Devuelve el primer elemento si la pila no está vacía
            return str(self.first)
        
        except IndexError:
            # Excepción de pila vacía
            return "La pila está vacía"
        
    
    def empty(self):
        """
        Comprueba si la pila está vacía.
        """
        if self.first == None:  # Pila vacía = True
            return True
        else:
            return False        # Pila no vacía = False
    
    def clear(self):
        """
        Elimina todos los elementos la pila.
        """
        self.first = None     # Actualizamos el puntero "first" a None.
        return "La pila se ha vaciado"


    


<h3> 1.1. Prueba de los métodos </h3>

In [12]:
# 1. Inicializamos la pila vacía
pila1 = LinkedStack() 
print("\n========= INICIALIZACIÓN DE PILA ===========")
print("Pila: ", pila1)

# 2. Añadimos elementos a la pila (Last-in)
print("\n========= MÉTODO PUSH (APILAR) ===========")
for num in range (1,10):
    pila1.push(num)
    print("Se ha apilado el elemento: ", num)
print("Pila: ", pila1)

# 3. Desapilamos la pila (First-out)
print("\n========= MÉTODO POP (DESAPILAR) ===========")
print("Se ha desapilado el elemento: ", pila1.pop(),)
print("Pila: ", pila1)

# 4. Identificamos la cima de la pila (Primer elemento)
print("\n========= MÉTODO PEEK (CIMA) ===========")
print("Cima: ", pila1.peek())
print("Pila: ", pila1)

# 5. Vaciamos la pila
print("\n========= MÉTODO CLEAR (LIMPIAR) ===========")
print("¿Está la pila vacía?: ", pila1.empty())
print("Aplicamos el método clear con éxito: ", pila1.clear())
print("Pila: ", pila1)

# 6. Comprobación de pila vacía
print("\n========= MÉTODO EMPTY (VACÍA) ===========")
print("¿Está la pila vacía?: ", pila1.empty())




========= INICIALIZACIÓN DE PILA ===========
Pila:  Pila vacía

========= MÉTODO PUSH (APILAR) ===========
Se ha apilado el elemento:  1
Se ha apilado el elemento:  2
Se ha apilado el elemento:  3
Se ha apilado el elemento:  4
Se ha apilado el elemento:  5
Se ha apilado el elemento:  6
Se ha apilado el elemento:  7
Se ha apilado el elemento:  8
Se ha apilado el elemento:  9
Pila:  ['9', '8', '7', '6', '5', '4', '3', '2', '1']

========= MÉTODO POP (DESAPILAR) ===========
Se ha desapilado el elemento:  9
Pila:  ['8', '7', '6', '5', '4', '3', '2', '1']

========= MÉTODO PEEK (CIMA) ===========
Cima:  8
Pila:  ['8', '7', '6', '5', '4', '3', '2', '1']

========= MÉTODO CLEAR (LIMPIAR) ===========
¿Está la pila vacía?:  False
Aplicamos el método clear con éxito:  La pila se ha vaciado
Pila:  Pila vacía

========= MÉTODO EMPTY (VACÍA) ===========
¿Está la pila vacía?:  True


<h2> 2. Pila dinámica con soporte Lista </h2>

In [14]:
# Creamos la clase

class ListStack:

    def __init__(self):
        """
        Inicializa una pila vacía con soporte de lista.
        """   
        self.pila = []  # Pila vacía
        
    def __str__(self):
        """
        Devuelve la pila como el string de la lista, en el orden LIFO.
        """   
        return str(self.pila) 

    def push(self, e):
        """
        Apila un elemento (e) al principio de la pila.
        """    
        self.pila.insert(0, e)  # Insertamos "e" en la posición 0 (inicio)
        return self.pila
    
    def pop(self):
        """
        Elimina y devuelve el 1º elemento de la pila.
        """    
        try:
            return self.pila.pop(0)        # Elimina y devuelve el 1º elemento de la lista
        except IndexError:                 # Excepción de pila vacía
            return "La pila está vacía"
            
    def peek(self):
        """
        Devuelve la cima de la pila (el 1º elemento).
        """
        try:
            return self.pila[0]           # Devuelve el 1º elemento de la lista
        except IndexError:                # Excepción de pila vacía
            return "La pila está vacía"
        
    def empty(self):
        """
        Comprueba si la pila está vacía.
        """
        if self.pila == []:               # Pila vacía = True
            return True
        else:                             # Pila no vacía = False
            return False
        
    def clear(self):
        """
        Elimina todos los elementos de la pila.
        """   
        self.pila.clear()
        return "La pila se ha vaciado"


<h3> 2.1. Prueba de los métodos </h3>

In [15]:
# 1. Inicializamos la pila vacía
pila2 = ListStack() 
print("\n========= INICIALIZACIÓN DE PILA ===========")
print("Pila: ", pila2)

# 2. Añadimos elementos a la pila (Last-in)
print("\n========= MÉTODO PUSH (APILAR) ===========")
for num in range (1,10):
    pila2.push(num)
    print("Se ha apilado el elemento: ", num)
print("Pila: ", pila2)

# 3. Desapilamos la pila (First-out)
print("\n========= MÉTODO POP (DESAPILAR) ===========")
print("Se ha desapilado el elemento: ", pila2.pop(),)
print("Pila: ", pila2)

# 4. Identificamos la cima de la pila (Primer elemento)
print("\n========= MÉTODO PEEK (CIMA) ===========")
print("Cima: ", pila2.peek())
print("Pila: ", pila2)

# 5. Vaciamos la pila
print("\n========= MÉTODO CLEAR (LIMPIAR) ===========")
print("¿Está la pila vacía?: ", pila2.empty())
print("Aplicamos el método clear con éxito: ", pila2.clear())
print("Pila: ", pila2)

# 6. Comprobación de pila vacía
print("\n========= MÉTODO EMPTY (VACÍA) ===========")
print("¿Está la pila vacía?: ", pila2.empty())




========= INICIALIZACIÓN DE PILA ===========
Pila:  []

========= MÉTODO PUSH (APILAR) ===========
Se ha apilado el elemento:  1
Se ha apilado el elemento:  2
Se ha apilado el elemento:  3
Se ha apilado el elemento:  4
Se ha apilado el elemento:  5
Se ha apilado el elemento:  6
Se ha apilado el elemento:  7
Se ha apilado el elemento:  8
Se ha apilado el elemento:  9
Pila:  [9, 8, 7, 6, 5, 4, 3, 2, 1]

========= MÉTODO POP (DESAPILAR) ===========
Se ha desapilado el elemento:  9
Pila:  [8, 7, 6, 5, 4, 3, 2, 1]

========= MÉTODO PEEK (CIMA) ===========
Cima:  8
Pila:  [8, 7, 6, 5, 4, 3, 2, 1]

========= MÉTODO CLEAR (LIMPIAR) ===========
¿Está la pila vacía?:  False
Aplicamos el método clear con éxito:  La pila se ha vaciado
Pila:  []

========= MÉTODO EMPTY (VACÍA) ===========
¿Está la pila vacía?:  True


<h2> 3. Pila dinámica con DeQue </h2>

In [10]:
# Usamos de soporte collections.deque
from collections import deque as dq

In [9]:
# Creamos la clase

class DqueStack:

    def __init__(self):
        """
        Inicializa una pila vacía con la función "deque".
        """   
        self.pila = dq()
        
    def __str__(self):
        """
        Devuelve la pila como el string de una lista.
        """   
        return str(list(self.pila))

    def push(self, e):
        """
        Apila un elemento (e) al principio de la pila.
        """    
        self.pila.appendleft(e)
        return self.pila
    
    def pop(self):
        """
        Elimina y devuelve el primer elemento de la pila.
        """    
        try:
            return self.pila.popleft()   # Elimina y devuelve el 1º elemento de la pila
        except IndexError:               # Excepción de pila vacía
            return "La pila está vacía"
            
    def peek(self):
        """
        Devuelve la cima de la pila (el primer elemento).
        """
        try:
            return self.pila[0]           # Devuelve el 1º elemento de la pila
        except IndexError:                # Excepción de pila vacía
            return "La pila está vacía"
        
    def empty(self):
        """
        Comprueba si la pila está vacía.
        """
        if self.pila == dq():             # Pila vacía = True
            return True
        else:                             # Pila no vacía = False
            return False
        
    def clear(self):
        """
        Elimina todos los elementos de la pila.
        Devuelve una pila vacía en forma de lista.
        """   
        self.pila.clear()
        return "La pila se ha vaciado"


<h3> 3.1. Prueba de los métodos </h3>

In [18]:
# 1. Inicializamos la pila vacía
pila3 = DqueStack() 
print("\n========= INICIALIZACIÓN DE PILA ===========")
print("Pila: ", pila3)

# 2. Añadimos elementos a la pila (Last-in)
print("\n========= MÉTODO PUSH (APILAR) ===========")
for num in range (1,10):
    pila3.push(num)
    print("Se ha apilado el elemento: ", num)
print("Pila: ", pila3)

# 3. Desapilamos la pila (First-out)
print("\n========= MÉTODO POP (DESAPILAR) ===========")
print("Se ha desapilado el elemento: ", pila3.pop(),)
print("Pila: ", pila3)

# 4. Identificamos la cima de la pila (Primer elemento)
print("\n========= MÉTODO PEEK (CIMA) ===========")
print("Cima: ", pila3.peek())
print("Pila: ", pila3)

# 5. Vaciamos la pila
print("\n========= MÉTODO CLEAR (LIMPIAR) ===========")
print("¿Está la pila vacía?: ", pila3.empty())
print("Aplicamos el método clear con éxito: ", pila3.clear())
print("Pila: ", pila3)

# 6. Comprobación de pila vacía
print("\n========= MÉTODO EMPTY (VACÍA) ===========")
print("¿Está la pila vacía?: ", pila3.empty())





========= INICIALIZACIÓN DE PILA ===========
Pila:  []

========= MÉTODO PUSH (APILAR) ===========
Se ha apilado el elemento:  1
Se ha apilado el elemento:  2
Se ha apilado el elemento:  3
Se ha apilado el elemento:  4
Se ha apilado el elemento:  5
Se ha apilado el elemento:  6
Se ha apilado el elemento:  7
Se ha apilado el elemento:  8
Se ha apilado el elemento:  9
Pila:  [9, 8, 7, 6, 5, 4, 3, 2, 1]

========= MÉTODO POP (DESAPILAR) ===========
Se ha desapilado el elemento:  9
Pila:  [8, 7, 6, 5, 4, 3, 2, 1]

========= MÉTODO PEEK (CIMA) ===========
Cima:  8
Pila:  [8, 7, 6, 5, 4, 3, 2, 1]

========= MÉTODO CLEAR (LIMPIAR) ===========
¿Está la pila vacía?:  False
Aplicamos el método clear con éxito:  La pila se ha vaciado
Pila:  []

========= MÉTODO EMPTY (VACÍA) ===========
¿Está la pila vacía?:  True


<h2> 4. Pila dinámica con soporte LifoQueue </h2>

In [19]:
# Usamos queue.LifoQueue de soporte
from queue import LifoQueue as lq

In [20]:
# Creamos la clase
class LifoQueueStack:
    
    def __init__(self):
        """
        Método constructor.
        Inicializa pila vacía con función queue.LifoQueue().
        """
        self.pila = lq()

    def __str__(self):
        """
        Devuelve la pila como el string de una lista.
        """
        pila_aux = []  # Crea una lista auxiliar

        if self.pila.empty():
            return "Pila vacía"

        while not self.pila.empty():
            elemento = self.pila.get()  # Variable para almacenar los elementos
            pila_aux.append(elemento)  # Almacena los elementos en pila_aux

    # Reinsertar elementos en la pila original
        for elemento in pila_aux[::-1]:
            self.pila.put(elemento)

        return str(pila_aux)


    def push(self, e):
        """
        Apila un elemento (e) al inicio de la pila.
        """
        self.pila.put(e)                       
        return self.pila
    
    def pop(self):
        """
        Extrae y muestra el primer elemento de la pila.
        """
        try:
            return self.pila.get()   
        except IndexError:                  # Si la pila está vacía, lanza una excepción
            return "La pila está vacía"
            
    def peek(self):
        """
        Devuelve la cima de la pila (el 1º elemento).
        """
        cima = self.pila.get()              # Guarda y elimina el 1º elemento en la variable "cima"
        self.pila.put(cima)                 # Reinsertamos el elemento
        try:
            return cima                         # Devuelve "cima"
        except IndexError:
            return "La pila está vacía"         # Si la pila está vacía, lanza una excepción
        
    def empty(self):
        """
        Comprueba si la pila está vacía.
        """
        if self.pila.empty():                   # Pila vacía = True
            return True
        else:                                   # Pila no vacía = False
            return False


    def clear(self):
        """
        Elimina todos los elementos de la pila.
        """
        while not self.pila.empty():
            self.pila.get()               # Mientras no esté vacía, elimina los elementos 1 a 1
        return "La pila se ha vaciado"

        


<h3> 4.1. Prueba de los métodos </h3>

In [21]:
# 1. Inicializamos la pila vacía
pila4 = LifoQueueStack() 
print("\n========= INICIALIZACIÓN DE PILA ===========")
print("Pila: ", pila4)

# 2. Añadimos elementos a la pila (Last-in)
print("\n========= MÉTODO PUSH (APILAR) ===========")
for num in range (1,10):
    pila4.push(num)
    print("Se ha apilado el elemento: ", num)
print("Pila: ", pila4)

# 3. Desapilamos la pila (First-out)
print("\n========= MÉTODO POP (DESAPILAR) ===========")
print("Se ha desapilado el elemento: ", pila4.pop(),)
print("Pila: ", pila4)

# 4. Identificamos la cima de la pila (Primer elemento)
print("\n========= MÉTODO PEEK (CIMA) ===========")
print("Cima: ", pila4.peek())
print("Pila: ", pila4)

# 5. Vaciamos la pila
print("\n========= MÉTODO CLEAR (LIMPIAR) ===========")
print("¿Está la pila vacía?: ", pila4.empty())
print("Aplicamos el método clear con éxito: ", pila4.clear())
print("Pila: ", pila4)

# 6. Comprobación de pila vacía
print("\n========= MÉTODO EMPTY (VACÍA) ===========")
print("¿Está la pila vacía?: ", pila4.empty())



========= INICIALIZACIÓN DE PILA ===========
Pila:  Pila vacía

========= MÉTODO PUSH (APILAR) ===========
Se ha apilado el elemento:  1
Se ha apilado el elemento:  2
Se ha apilado el elemento:  3
Se ha apilado el elemento:  4
Se ha apilado el elemento:  5
Se ha apilado el elemento:  6
Se ha apilado el elemento:  7
Se ha apilado el elemento:  8
Se ha apilado el elemento:  9
Pila:  [9, 8, 7, 6, 5, 4, 3, 2, 1]

========= MÉTODO POP (DESAPILAR) ===========
Se ha desapilado el elemento:  9
Pila:  [8, 7, 6, 5, 4, 3, 2, 1]

========= MÉTODO PEEK (CIMA) ===========
Cima:  8
Pila:  [8, 7, 6, 5, 4, 3, 2, 1]

========= MÉTODO CLEAR (LIMPIAR) ===========
¿Está la pila vacía?:  False
Aplicamos el método clear con éxito:  La pila se ha vaciado
Pila:  Pila vacía

========= MÉTODO EMPTY (VACÍA) ===========
¿Está la pila vacía?:  True


<h2> Comparativa de métodos </h2>

Los distintos constructores que hemos empleado en cada método son:

- Clase `LinkedStack`

    Creamos un puntero "first": `self.first = None`

- Clase `ListStack`

    Creamos una lista para almacenar los elementos de la pila: `self.pila = []`
    
- Clase `DqueStack`

    Creamos una pila vacía con la función queue.deque(): `self.pila = dq()`

- Clase `LifoQueueStack`

    Creamos una pila vacía con la función queue.LifoQueue(): `self.pila = lq()`

**Facilidad de uso y comprensión**: 

- Mediante el sistema de lista enlazada, hay que actualizar el puntero, lo que hace que el código sea más elaborado y menos fácil de comprender. En lugar de utilizar funciones útiles para la implementación de los métodos, tenemos que trabajar con sentencias "if" y haciendo uso de otra clase auxiliar Nodo(), lo que no es nada eficiente.

- Usando de soporte una lista de Python podemos iterar, lo que hace más fácil almacenar los elementos de la pila. Además, podemos hacer uso de funciones útiles para implementar en los métodos de la clase:

    - insert(index, e): Nos permite insertar un elemento (e) en el inicio de la pila (índice=0)
    - pop(index): Nos permite extraer y devolver el elemento del inicio de la pila (índice=0)

- Con queue.deque() tenemos funciones que facilitan el uso y comprensión del código:

    - appendleft(e): Nos permite apilar un elemento al inicio de la pila
    - popleft(): Nos permite extraer y devolver el elemento del inicio de la pila

    Además, hace que escribamos el código en menos líneas.
    
- Con queue.LifoQueue() también tenemos funciones específicas para pilas:

    - put(e): Nos permite apilar un elemento al inicio de la pila
    - get(): Nos permite extraer y devolver el elemento del inicio de la pila

    Aun así, no posee una función para mostrar elementos, ya que solo elimina el elemento inicial con get(). Por ello, hubo que hacer más uso de líneas de código almacenando el elemento, para luego reintroducirlo en la pila y mostrarlo mediante peek().

Después de implementar los 6 métodos en las diferentes clases, llego a la conclusión de que el código más fácil de usar y comprender es el de DqueStack gracias a las funciones de queue.deque. 

No es así con queue.LifoQueue ya que sus funciones son más limitadas, además de que no hay manera de iterar en la pila.




<h2> Comparación de rendimiento </h2>

Compararemos la rapidez del código para cada una de las 4 clases haciendo uso de `%%timeit`

<h3> 1) LinkedStack </h3>

In [23]:
%%timeit

pila1 = LinkedStack()
for num in range(1,100):
    pila1.push(num)
pila1.peek()
pila1.pop
pila1.clear
pila1.empty

14.8 µs ± 517 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


<h3> 2) ListStack </h3>

In [27]:
%%timeit

pila2 = ListStack()
for num in range(1,100):
    pila2.push(num)
pila2.peek()
pila2.pop
pila2.clear
pila2.empty

8.51 µs ± 521 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


<h3> 2) DqueStack </h3>

In [25]:
%%timeit

pila3 = DqueStack()
for num in range(1,100):
    pila3.push(num)
pila3.peek()
pila3.pop
pila3.clear
pila3.empty

5.65 µs ± 151 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


<h3> 4) LifoQueueStack </h3>

In [26]:
%%timeit

pila4 = LifoQueueStack()
for num in range(1,100):
    pila4.push(num)
pila4.peek()
pila4.pop
pila4.clear
pila4.empty

59.3 µs ± 1.38 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


<h3> Resultados</h3>

- `DqueStack`: La implementación DqueStack es la más rápida con un tiempo promedio de ejecución de 5.65 microsegundos por loop.

- `ListStack`: La implementación ListStack se ubica en segundo lugar con un tiempo promedio de ejecución de 8.51 microsegundos por loop.

- `LinkedStack`: La implementación LinkedStack es un poco más lenta que ListStack con un tiempo promedio de ejecución de 14.8 microsegundos por loop.

- `LifoQueueStack`: La implementación LifoQueueStack es la más lenta con un tiempo promedio de ejecución de 59.3 microsegundos por loop (y solo se ejecutaron 10,000 loops en la medición).

<h3> Conclusiones </h3>

DqueStack aprovecha la eficiencia de collections.deque para realizar operaciones de push y pop de forma rápida.

ListStack utiliza listas de Python, que son eficientes para la inserción y eliminación al final (operaciones LIFO de la pila).

LinkedStack implica la creación y manipulación de nodos enlazados, lo que puede ser ligeramente más lento que las estructuras de datos basadas en arrays.

LifoQueueStack parece tener una implementación menos eficiente para las operaciones de pila, ya que su tiempo de ejecución es significativamente mayor y se ejecutaron menos loops en la medición.

In [19]:
Piiiiiiiiiiiiii = DqueStack()

Piiiiiiiiiiiiii.push("Viva el Sevilla")
Piiiiiiiiiiiiii.push(21)
Piiiiiiiiiiiiii.push("pipi")

print(Piiiiiiiiiiiiii)

Piiiiiiiiiiiiii.pop()
print(Piiiiiiiiiiiiii)
Piiiiiiiiiiiiii.peek()
Piiiiiiiiiiiiii.clear()
print(Piiiiiiiiiiiiii)


['pipi', 21, 'Viva el Sevilla']
[21, 'Viva el Sevilla']
[]
